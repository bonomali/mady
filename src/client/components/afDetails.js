// @flow

import React from 'react';
import { graphql } from 'react-relay';
import moment from 'moment';
import { flexItem, Icon, LargeMessage } from 'giu';
import _t from '../../translate';
import { COLORS } from '../gral/constants';
import QueryRendererWrapper from './uuQueryRendererWrapper';

// ==========================================
// Component declarations
// ==========================================
type Props = {
  // lang: string,
  selectedKeyId: ?string,
};

const query = graphql`
  query afDetailsQuery($selectedKeyId: ID!) {
    node(id: $selectedKeyId) {
      ... on Key {
        firstUsed
        unusedSince
        description
        sources
      }
    }
  }
`;

// ==========================================
// Component
// ==========================================
class Details extends React.PureComponent {
  props: Props;

  constructor(props) {
    super(props);
    this.state = {
      key: props.node,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.node) this.setState({ key: nextProps.node });
  }

  render() {
    this.key = this.props.node;
    return (
      <div style={style.outer}>
        <div style={style.title}>
          {_t('msgDetailsView_Details').toUpperCase()}
        </div>
        {this.renderContents()}
      </div>
    );
  }

  renderContents() {
    if (this.props.selectedKeyId == null) {
      return (
        <LargeMessage>
          {_t('msgDetailsView_No message selected')}
        </LargeMessage>
      );
    }
    if (!this.state.key) {
      return (
        <LargeMessage>
          <Icon icon="circle-o-notch" />
        </LargeMessage>
      );
    }
    const { description, sources, firstUsed, unusedSince } = this.state.key;
    const since = this.renderDate(firstUsed);
    const until = unusedSince
      ? <span>
          {' '}{_t('msgDetailsView_until')} {this.renderDate(unusedSince)}
        </span>
      : ':';
    const elSources = sources.length
      ? <ul style={style.srcList}>
          {sources.map((src, idx) =>
            <li key={idx}>
              {src}
            </li>
          )}
        </ul>
      : null;
    return (
      <div>
        {description &&
          <div>
            {description}
          </div>}
        {_t('msgDetailsView_Used since')} {since}
        {until}
        {elSources}
      </div>
    );
  }

  renderDate(d: string) {
    return (
      <span title={moment(d).format('LLLL')} style={style.date}>
        {moment(d).fromNow()}
      </span>
    );
  }
}

// ------------------------------------------
const style = {
  outer: flexItem('none', {
    minHeight: 110,
    backgroundColor: COLORS.medium,
    padding: 5,
    marginTop: 5,
  }),
  title: {
    fontWeight: 900,
    letterSpacing: 3,
    textAlign: 'center',
    marginBottom: 10,
  },
  date: {
    fontWeight: 'bold',
    color: '#444',
  },
  srcList: {
    marginTop: 0,
  },
};

// ==========================================
// Public API
// ==========================================
const Container = ({ selectedKeyId, ...otherProps }) =>
  <QueryRendererWrapper
    query={query}
    vars={{ selectedKeyId }}
    Component={Details}
    renderDuringLoad
    selectedKeyId={selectedKeyId}
    {...otherProps}
  />;
export default Container;
export { Details as _Details };
