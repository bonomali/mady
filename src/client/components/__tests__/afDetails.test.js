/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import { _Details as Details } from '../afDetails';

// https://github.com/facebook/react/issues/7386#issuecomment-238091398
jest.mock('react-dom');

// ======================================================
// Fixtures
// ======================================================
const KEY_ID = 'keyId';

const VIEWER_WITHOUT_KEY = {
  id: 'me',
};

const VIEWER_WITH_KEY_USED = {
  id: 'me',
  anyNode: {
    firstUsed: '2016-04-18T05:37:47.074Z',
    unusedSince: null,
    description: 'Some details',
    sources: ['src/client/components/060-translator.js'],
  },
};

const VIEWER_WITH_KEY_UNUSED = {
  id: 'me',
  anyNode: {
    firstUsed: '2016-04-18T05:37:47.074Z',
    unusedSince: '2016-04-21T05:37:47.074Z',
    description: 'Some details',
    sources: [],
  },
};

const RELAY_MOCK = {
  setVariables: jest.fn(),
};

// ======================================================
// Tests
// ======================================================
describe('Details', () => {
  it('renders correctly without a selection', () => {
    const tree = renderer
      .create(
        <Details
          relay={RELAY_MOCK}
          viewer={VIEWER_WITHOUT_KEY}
          selectedKeyId={null}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with a selection but no data yet', () => {
    const tree = renderer
      .create(
        <Details
          relay={RELAY_MOCK}
          viewer={VIEWER_WITHOUT_KEY}
          selectedKeyId={KEY_ID}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('when the details for the selection are available', () => {
    it('renders correctly for a message that is used', () => {
      const tree = renderer
        .create(
          <Details
            relay={RELAY_MOCK}
            viewer={VIEWER_WITH_KEY_USED}
            selectedKeyId={KEY_ID}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders correctly for a message that is unused', () => {
      const tree = renderer
        .create(
          <Details
            relay={RELAY_MOCK}
            viewer={VIEWER_WITH_KEY_UNUSED}
            selectedKeyId={KEY_ID}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
