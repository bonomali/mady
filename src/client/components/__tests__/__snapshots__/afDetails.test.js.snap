// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details renders correctly with a selection but no data yet 1`] = `
<div
  style={
    Object {
      "WebkitFlex": "none",
      "backgroundColor": "#dcd6ff",
      "flex": "none",
      "marginTop": 5,
      "minHeight": 110,
      "padding": 5,
    }
  }
>
  <div
    style={
      Object {
        "fontWeight": 900,
        "letterSpacing": 3,
        "marginBottom": 10,
        "textAlign": "center",
      }
    }
  >
    DETAILS
  </div>
  <div
    className="giu-large-message"
    style={
      Object {
        "color": "#999999",
        "fontSize": "1.4em",
        "fontWeight": 700,
        "padding": "0.8em",
        "textAlign": "center",
      }
    }
  >
    <i
      className="giu-icon fa fa-circle-o-notch fa-spin"
      style={
        Object {
          "color": undefined,
          "cursor": "pointer",
          "fontSize": undefined,
          "letterSpacing": "normal",
        }
      }
    />
  </div>
</div>
------------HTML PREVIEW---------------
<div style="-webkit-flex:none;background-color:#dcd6ff;flex:none;margin-top:5px;min-height:110px;padding:5px;">
  <div style="font-weight:900;letter-spacing:3px;margin-bottom:10px;text-align:center;">
    DETAILS
  </div>
  <div
    class="giu-large-message"
    style="color:#999999;font-size:1.4em;font-weight:700;padding:0.8em;text-align:center;"
  >
    <i
      class="giu-icon fa fa-circle-o-notch fa-spin"
      style="cursor:pointer;letter-spacing:normal;"
    >
    </i>
  </div>
</div>
`;

exports[`Details renders correctly without a selection 1`] = `
<div
  style={
    Object {
      "WebkitFlex": "none",
      "backgroundColor": "#dcd6ff",
      "flex": "none",
      "marginTop": 5,
      "minHeight": 110,
      "padding": 5,
    }
  }
>
  <div
    style={
      Object {
        "fontWeight": 900,
        "letterSpacing": 3,
        "marginBottom": 10,
        "textAlign": "center",
      }
    }
  >
    DETAILS
  </div>
  <div
    className="giu-large-message"
    style={
      Object {
        "color": "#999999",
        "fontSize": "1.4em",
        "fontWeight": 700,
        "padding": "0.8em",
        "textAlign": "center",
      }
    }
  >
    No message selected
  </div>
</div>
------------HTML PREVIEW---------------
<div style="-webkit-flex:none;background-color:#dcd6ff;flex:none;margin-top:5px;min-height:110px;padding:5px;">
  <div style="font-weight:900;letter-spacing:3px;margin-bottom:10px;text-align:center;">
    DETAILS
  </div>
  <div
    class="giu-large-message"
    style="color:#999999;font-size:1.4em;font-weight:700;padding:0.8em;text-align:center;"
  >
    No message selected
  </div>
</div>
`;

exports[`Details when the details for the selection are available renders correctly for a message that is unused 1`] = `
<div
  style={
    Object {
      "WebkitFlex": "none",
      "backgroundColor": "#dcd6ff",
      "flex": "none",
      "marginTop": 5,
      "minHeight": 110,
      "padding": 5,
    }
  }
>
  <div
    style={
      Object {
        "fontWeight": 900,
        "letterSpacing": 3,
        "marginBottom": 10,
        "textAlign": "center",
      }
    }
  >
    DETAILS
  </div>
  <div
    className="giu-large-message"
    style={
      Object {
        "color": "#999999",
        "fontSize": "1.4em",
        "fontWeight": 700,
        "padding": "0.8em",
        "textAlign": "center",
      }
    }
  >
    <i
      className="giu-icon fa fa-circle-o-notch fa-spin"
      style={
        Object {
          "color": undefined,
          "cursor": "pointer",
          "fontSize": undefined,
          "letterSpacing": "normal",
        }
      }
    />
  </div>
</div>
------------HTML PREVIEW---------------
<div style="-webkit-flex:none;background-color:#dcd6ff;flex:none;margin-top:5px;min-height:110px;padding:5px;">
  <div style="font-weight:900;letter-spacing:3px;margin-bottom:10px;text-align:center;">
    DETAILS
  </div>
  <div
    class="giu-large-message"
    style="color:#999999;font-size:1.4em;font-weight:700;padding:0.8em;text-align:center;"
  >
    <i
      class="giu-icon fa fa-circle-o-notch fa-spin"
      style="cursor:pointer;letter-spacing:normal;"
    >
    </i>
  </div>
</div>
`;

exports[`Details when the details for the selection are available renders correctly for a message that is used 1`] = `
<div
  style={
    Object {
      "WebkitFlex": "none",
      "backgroundColor": "#dcd6ff",
      "flex": "none",
      "marginTop": 5,
      "minHeight": 110,
      "padding": 5,
    }
  }
>
  <div
    style={
      Object {
        "fontWeight": 900,
        "letterSpacing": 3,
        "marginBottom": 10,
        "textAlign": "center",
      }
    }
  >
    DETAILS
  </div>
  <div
    className="giu-large-message"
    style={
      Object {
        "color": "#999999",
        "fontSize": "1.4em",
        "fontWeight": 700,
        "padding": "0.8em",
        "textAlign": "center",
      }
    }
  >
    <i
      className="giu-icon fa fa-circle-o-notch fa-spin"
      style={
        Object {
          "color": undefined,
          "cursor": "pointer",
          "fontSize": undefined,
          "letterSpacing": "normal",
        }
      }
    />
  </div>
</div>
------------HTML PREVIEW---------------
<div style="-webkit-flex:none;background-color:#dcd6ff;flex:none;margin-top:5px;min-height:110px;padding:5px;">
  <div style="font-weight:900;letter-spacing:3px;margin-bottom:10px;text-align:center;">
    DETAILS
  </div>
  <div
    class="giu-large-message"
    style="color:#999999;font-size:1.4em;font-weight:700;padding:0.8em;text-align:center;"
  >
    <i
      class="giu-icon fa fa-circle-o-notch fa-spin"
      style="cursor:pointer;letter-spacing:normal;"
    >
    </i>
  </div>
</div>
`;
